plugin/access.lua

```diff
@@ -14,170 +37,223 @@ local function array_has_value(arr, val)
 end
 
 --
--- Check for a required scope if configured
+-- Optionally check scopes configured for a location
 --
 local function verify_scope(jwt_text, required_scope)
     
     if required_scope == nil then
         return true
     end
 
-    local data, err = jwt_parser:new(jwt_text)
-    if err then
-        ngx.log(ngx.WARN, "Unable to parse JWT access token: " .. err)
+    local data = jwt:load_jwt(jwt_text, nil)
+    if not data.valid then
+        local details = 'Unable to parse JWT access token'
+        if data.reason then
+            details = details .. ': ' .. data.reason
+        end
+        ngx.log(ngx.WARN, details)
         return false
     end
 
-    local scope_values = pl_stringx.split(data.claims.scope, " ")
-    for id, scope in ipairs(scope_values) do
-        if scope == required_scope then
-            return true
+    if not data.payload.scope then
+        return false
+    end
+
+    local required_scope_parts = string.gmatch(required_scope, "%S+")
+    local actual_scope_parts   = iterator_to_array(string.gmatch(data.payload.scope, "%S+"))
+
+    for required_value in required_scope_parts do
+        if not array_has_value(actual_scope_parts, required_value) then
+            ngx.log(ngx.WARN, 'The required scope ' .. required_value .. ' was not found in the received access token')
+            return false
         end
     end
 
-    return false
+    return true
 end
 
 --
--- Get the access token from the cache or introspect the token if not found
--- We use the access_token value as the cache key
--- Note also that we pass two arguments to the introspect_access_token callback
+-- Get the token from the cache or introspect it
 --
 local function verify_access_token(access_token, config)
-    
-    local res, err = kong.cache:get(
-        access_token,
-        { ttl = config.token_cache_seconds },
-        introspect_access_token,
-        access_token,
-        config)
-
-    if err then
-        local cacheMessage = "A technical problem occurred during cache access"
-        ngx.log(ngx.WARN, cacheMessage .. err)
-        error_response(ngx.HTTP_INTERNAL_SERVER_ERROR, "cache_error", cacheMessage, config)
-    end 
 
-    if res.status ~= 200 then
-        kong.cache:invalidate(access_token)
+    -- Return previous introspeciton results for the same token if available
+    local dict = ngx.shared['phantom-token']
+    local existing_jwt = dict:get(access_token)
+    if existing_jwt then
+        return { status = 200, jwt = existing_jwt }
     end
 
-    if res.status == 204 then
-        ngx.log(ngx.WARN, "Received a " .. res.status .. " introspection response due to the access token being invalid or expired")
-        invalid_token_error_response(config)
-    end
+    -- Otherwise introspect the opaque access token
+    local result = introspect_access_token(access_token, config)
+    if result.status == 200 then
 
-    if res.status ~= 200 then
+        if not verify_scope(result.jwt, config.scope) then
```

Comment:
I think it makes sense to cache the result of verification. Otherwise, we will repeatedly introspect tokens that do not have the proper scope. The scope of the opaque token can't change, right?

Reply:
That's a good point. I have fixed this by caching the introspection result and then checking the scope. As you say, this would reduce load on the authorization server.